src/main/java/org/opensearch/security/http/OnBehalfOfAuthenticator.java

```diff
@@ -0,0 +1,249 @@
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.security.http;
+
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.JwtParser;
+import io.jsonwebtoken.security.WeakKeyException;
+import org.apache.hc.core5.http.HttpHeaders;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import org.opensearch.OpenSearchException;
+import org.opensearch.OpenSearchSecurityException;
+import org.opensearch.SpecialPermission;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.rest.RestChannel;
+import org.opensearch.rest.RestRequest;
+import org.opensearch.security.auth.HTTPAuthenticator;
+import org.opensearch.security.authtoken.jwt.EncryptionDecryptionUtil;
+import org.opensearch.security.ssl.util.ExceptionUtils;
+import org.opensearch.security.user.AuthCredentials;
+import org.opensearch.security.util.KeyUtils;
+
+import static org.opensearch.security.OpenSearchSecurityPlugin.LEGACY_OPENDISTRO_PREFIX;
+import static org.opensearch.security.OpenSearchSecurityPlugin.PLUGINS_PREFIX;
+
+public class OnBehalfOfAuthenticator implements HTTPAuthenticator {
+
+    private static final String REGEX_PATH_PREFIX = "/(" + LEGACY_OPENDISTRO_PREFIX + "|" + PLUGINS_PREFIX + ")/" + "(.*)";
+    private static final Pattern PATTERN_PATH_PREFIX = Pattern.compile(REGEX_PATH_PREFIX);
+    private static final String ON_BEHALF_OF_SUFFIX = "api/generateobotoken";
+    private static final String ACCOUNT_SUFFIX = "api/account";
+
+    protected final Logger log = LogManager.getLogger(this.getClass());
+
+    private static final Pattern BEARER = Pattern.compile("^\\s*Bearer\\s.*", Pattern.CASE_INSENSITIVE);
+    private static final String BEARER_PREFIX = "bearer ";
+
+    private final JwtParser jwtParser;
+    private final String encryptionKey;
+    private final Boolean oboEnabled;
+    private final String clusterName;
+
+    public OnBehalfOfAuthenticator(Settings settings, String clusterName) {
+        String oboEnabledSetting = settings.get("enabled");
+        oboEnabled = oboEnabledSetting == null ? Boolean.TRUE : Boolean.valueOf(oboEnabledSetting);
+        encryptionKey = settings.get("encryption_key");
+        jwtParser = initParser(settings.get("signing_key"));
+        this.clusterName = clusterName;
+    }
+
+    private JwtParser initParser(final String signingKey) {
+        JwtParser _jwtParser = KeyUtils.createJwtParserFromSigningKey(signingKey, log);
+
+        if (_jwtParser == null) {
+            throw new RuntimeException("Unable to find on behalf of authenticator signing key");
```

Comment:
Curious, why is this done here opposed to being done within `KeyUtils.createJwtParserFromSigningKey`?

Reply:
I was actually thinking in implementing this logic in this way, but since I also merged some identical logic in `HTTPJwtAuthenticator` ([here](https://github.com/opensearch-project/security/pull/3179/files#diff-136c07af96944229a50e19e07ccae6ddf8a86bd4464b32f87a4930cac688fdfbL66)) too, I thought this msg was specific for obo authenticator.